solidity/contracts/Gravity.sol

```diff
@@ -425,21 +437,44 @@ function updateValset(
 				// Send transaction amounts to destinations
 				uint256 totalFee;
 				for (uint256 i = 0; i < _amounts.length; i++) {
-					IERC20(_tokenContract).safeTransfer(_destinations[i], _amounts[i]);
+					transferNoRevert(_tokenContract, _destinations[i], _amounts[i]);
 					totalFee = totalFee + _fees[i];
 				}
 
 				// Send transaction fees to msg.sender
-				IERC20(_tokenContract).safeTransfer(msg.sender, totalFee);
+				transferNoRevert(_tokenContract, msg.sender, totalFee);
 			}
 		}
-
 		// LOGS scoped to reduce stack depth
 		{
 			state_lastEventNonce = state_lastEventNonce + 1;
 			emit TransactionBatchExecutedEvent(_batchNonce, _tokenContract, state_lastEventNonce);
 		}
-	}	
+	}
+
+	function transferNoRevert(address token, address to, uint value) internal {
+		try IERC20(token).transfer(to, value) returns (bool success){
+			if (!success) {
+				state_RevertedVouchers[state_lastRevertedNonce] = TransferReverted(token, to, value, true);
+				state_lastRevertedNonce = state_lastRevertedNonce + 1;
+			}
+		} catch {
+			state_RevertedVouchers[state_lastRevertedNonce] = TransferReverted(token, to, value, true);
+			state_lastRevertedNonce = state_lastRevertedNonce + 1;
+		}
```

Comment:
Will there be any difference between a false return from `transfer` and error thrown? For example one thing in my mind is whether the transfer is reverted or is at a limbo state between the two.

Or is this not a concern?

Reply:
its a valid concern and the answer is that I am not sure.

I think transfer can return false when the amount exceed what sender balance. but in our case it should be impossible (crc20 supply = gravity balance), if it fails for this case it is most likely a bug in our side

Could it be other case of failure?

Reply:
The function `safeTransfer` ensures that the error would be raised.

Some example here:
revert
`transfer` returns `false` from ERC20
`transfer` throws error from ERC20

won't revert
`transfer` returns `true` from ERC20
`transfer` didn't return anything

One of the edge case @thomas-nguy suggested is the recipient address is blacklisted, which I agree.